### ISSUE TITLE
`cargo clippy --fix` bug on redundant_closure_call

### ISSUE THREAD

**Issue by explosion-mental** [rust-lang/rust#112968]
### Code

```Rust
// init directories
    let Some(config_path) = (|| -> Option<PathBuf> {
        if let Some(c) = &cli.config_path {
            Some(c.to_owned())
        } else {
            dirs::config_dir()
        }
    })() else {
        anyhow::bail!("Config path for the platform could not be found, please report this at <https://codeberg.org/explosion-mental/wallust/issues>");
    };
```

### Current output

```Shell
Checking wallust v2.4.1 (/path/to/wallust)
warning: failed to automatically apply fixes suggested by rustc to crate `wallust`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error: right curly brace `}` before `else` in a `let...else` statement not allowed
  --> src/main.rs:31:5
   |
31 |     } else {
   |     ^
   |
help: wrap the expression in parentheses
   |
25 ~     let Some(config_path) = ({
26 |         if let Some(c) = &cli.config_path {
 ...
30 |         }
31 ~     }) else {
   |

error: aborting due to previous error

Original diagnostics will follow.

warning: try not to call a closure in the expression where it is declared
  --> src/main.rs:25:29
   |
25 |       let Some(config_path) = (|| -> Option<PathBuf> {
   |  _____________________________^
26 | |         if let Some(c) = &cli.config_path {
27 | |             Some(c.to_owned())
28 | |         } else {
29 | |             dirs::config_dir()
30 | |         }
31 | |     })() else {
   | |________^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure_call
   = note: `#[warn(clippy::redundant_closure_call)]` on by default
help: try doing something like
   |
25 ~     let Some(config_path) = {
26 +         if let Some(c) = &cli.config_path {
27 +             Some(c.to_owned())
28 +         } else {
29 +             dirs::config_dir()
30 +         }
31 ~     } else {
   |
```

### Desired output

```Shell
Checking wallust v2.4.1 (/path/to/wallust)
       Fixed src/main.rs (1 fix)
    Finished dev [unoptimized + debuginfo] target(s) in X.XXs
```

### Rationale and extra context

Using `cargo clippy --fix` aborted with the above message, indicating that **could** be a cargo or rustc bug. 

Edit: fixed markdown

**Comment by compiler-errors** [comment]
What version of rust are you running? This really should be filed against clippy instead of rust, and I thought that this was fixed :thinking:

**Comment by explosion-mental** [comment]
>  This really should be filed against clippy instead of rust

I guess, I followed what the output said. If this belongs to clippy then feel free to close it, I'll move it there

**cargo**
```
cargo 1.69.0 (6e9a83356 2023-04-12)
```

**rustc**
```
rustc 1.69.0 (84c898d65 2023-04-16)
```

**clippy** `cargo clippy --version`
```
clippy 0.1.69 (84c898d 2023-04-16)
```

**Comment by compiler-errors** [comment]
Yeah, you should open this against clippy instead I think, and you may also want to upgrade to 1.70
